### PR TITLE
Account for possible failures during attribute retargeting.

### LINKED ISSUE #65048: Retargeting asserts when attribute forwarded in NoPia scenario
The basic scenario here is when:

1. An initial compilation occurs that involves NoPia and a reference to `System.Runtime.InteropServices.dll` which houses `PrimaryInteropAssemblyAttribute`. 
2. A subsequent compilation using the reference generated from 1, against a new set of references but `PrimaryInteropAssemblyAttribute` was forwarded to a new assembly. 

A working repro can be found by doing the following: 

- `git remote add jaredpar https://github.com/jaredpar/roslyn.git`
- `git fetch jaredpar`
- `git checkout -B repro jaredpar/repro-65048` 
- Execute `DefaultInterfaceImplementationTests.NoPia_09` on `net472` 

The retargeting layer cannot find the appropriate constructor as the assembly is missing. This triggers an assert in `SourceAttributeData` as it's expecting either a symbol or an error 

```csharp
Debug.Assert(attributeConstructor is object || hasErrors);
```

The stack trace of the assert is as follows

```
    ThrowingTraceListener.Fail(String message, String detailMessage) line 26
    TraceInternal.Fail(String message)
    Debug.Assert(Boolean condition)
    SourceAttributeData.ctor(SyntaxReference applicationNode, NamedTypeSymbol attributeClass, MethodSymbol attributeConstructor, ImmutableArray`1 constructorArguments, ImmutableArray`1 constructorArgumentsSourceIndices, ImmutableArray`1 namedArguments, Boolean hasErrors, Boolean isConditionallyOmitted) line 44
    RetargetingAttributeData.ctor(SyntaxReference applicationNode, NamedTypeSymbol attributeClass, MethodSymbol attributeConstructor, ImmutableArray`1 constructorArguments, ImmutableArray`1 constructorArgumentsSourceIndices, ImmutableArray`1 namedArguments, Boolean hasErrors, Boolean isConditionallyOmitted) line 24
    RetargetingSymbolTranslator.RetargetAttributeData(CSharpAttributeData oldAttributeData) line 1216
    <>c.<RetargetAttributes>b__45_0(CSharpAttributeData a, RetargetingSymbolTranslator t) line 1167
    ImmutableArray.CreateRange[TSource,TArg,TResult](ImmutableArray`1 items, Func`3 selector, TArg arg)
    ImmutableArrayExtensions.SelectAsArray[TItem,TArg,TResult](ImmutableArray`1 items, Func`3 map, TArg arg) line 156
    RetargetingSymbolTranslator.RetargetAttributes(ImmutableArray`1 oldAttributes) line 1167
    <5 more frames...>
    CSharpCompilation.GetSourceDeclarationDiagnostics(SyntaxTree syntaxTree, Nullable`1 filterSpanWithinTree, Func`4 locationFilterOpt, CancellationToken cancellationToken) line 2959
    CSharpCompilation.GetDiagnosticsWithoutFiltering(CompilationStage stage, Boolean includeEarlierStages, BindingDiagnosticBag builder, CancellationToken cancellationToken) line 2729
    CSharpCompilation.GetDiagnostics(CompilationStage stage, Boolean includeEarlierStages, DiagnosticBag diagnostics, CancellationToken cancellationToken) line 2655
    CSharpCompilation.GetDiagnostics(CompilationStage stage, Boolean includeEarlierStages, CancellationToken cancellationToken) line 2647
    CSharpCompilation.CompileMethods(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, Boolean emitMetadataOnly, Boolean emitTestCoverageData, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken) line 3201
    Compilation.Emit(Stream peStream, Stream metadataPEStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, RebuildData rebuildData, CompilationTestData testData, CancellationToken cancellationToken) line 2890
    CompilationExtensions.EmitToArray(Compilation compilation, EmitOptions options, CompilationTestData testData, DiagnosticDescription[] expectedWarnings, Stream pdbStream, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, IEnumerable`1 manifestResources, Stream metadataPEStream) line 74
    CompilationExtensions.EmitToPortableExecutableReference(Compilation comp, EmitOptions options, Boolean embedInteropTypes, ImmutableArray`1 aliases, DiagnosticDescription[] expectedWarnings) line 128
    CompilationExtensions.EmitToImageReference(Compilation comp, EmitOptions options, Boolean embedInteropTypes, ImmutableArray`1 aliases, DiagnosticDescription[] expectedWarnings) line 119
    DefaultInterfaceImplementationTests.NoPia_09() line 50583
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #65048.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.cs [hunk 1]
```diff
@@ -1220,7 +1220,7 @@ private CSharpAttributeData RetargetAttributeData(CSharpAttributeData oldAttribu
                     newAttributeCtorArguments,
                     oldAttribute.ConstructorArgumentsSourceIndices,
                     newAttributeNamedArguments,
-                    oldAttribute.HasErrors,
+                    oldAttribute.HasErrors || newAttributeCtor is null,
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.cs:1223]: Consider naming this argument for future clarity.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetCustomAttributes.cs [hunk 2]
```diff
@@ -321,5 +321,77 @@ public class C
...
+            var comp1 = CreateCompilation(source1, options: TestOptions.DebugDll);
+
+            var source2 = @"
+[A]
+public class C1
+{}
+";
+
+            var comp2 = CreateCompilation(source2, references: new[] { comp1.ToMetadataReference() }, options: TestOptions.DebugDll);
+            var comp3 = CreateCompilation("", references: new[] { comp2.ToMetadataReference() }, options: TestOptions.DebugDll);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetCustomAttributes.cs:346]: For my own edification: is the fact that `A` is missing here the reason we get into a retargeting scenario at all?

**Review by 333fred — APPROVED** [review]
A minor comment and a question, otherwise LGTM (commit 1)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetCustomAttributes.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetCustomAttributes.cs:346]: > For my own edification: is the fact that ```A``` is missing here the reason we get into a retargeting scenario at all?

To be precise, the fact that the reference for `comp1` is not provided is the reason for retargeting.

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.cs:1223]:
```suggestion
                    hasErrors: oldAttribute.HasErrors || newAttributeCtor is null,
```

File: src/Compilers/VisualBasic/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.vb [hunk 3]
```diff
@@ -699,7 +699,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Retargeting
                                                     newCtorArguments,
                                                     newNamedArguments,
                                                     oldAttribute.IsConditionallyOmitted,
-                                                    oldAttribute.HasErrors)
+                                                    oldAttribute.HasErrors OrElse newAttributeCtor Is Nothing)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/Retargeting/RetargetingSymbolTranslator.vb:702]:
```suggestion
                                                    hasErrors:=oldAttribute.HasErrors OrElse newAttributeCtor Is Nothing)
```

**Pull request merged**